protocols/identify/src/lib.rs

```diff
@@ -229,15 +248,52 @@ impl<T: AddrManager> ServiceProtocol for IdentifyProtocol<T> {
                     }
                 } else {
                     trace!("received observed address: {}", addr);
-                    info.observed_addr = Some(addr.clone());
-                    // TODO how can we trust this address?
-                    if self
-                        .addr_mgr
-                        .add_observed_addr(&info.peer_id, addr.clone())
-                        .is_disconnect()
-                    {
-                        service.disconnect(session.id);
+
+                    // TODO: this address shoulde be reachable
+                    // Add transform observed address to local listen address list
+                    if let Some(socket_addr) = multiaddr_to_socketaddr(&addr) {
+                        let observed_ip = socket_addr.ip();
+
+                        // replace observed address's port part
+                        if self
+                            .listen_addrs
+                            .iter()
+                            .filter_map(|listen_addr| multiaddr_to_socketaddr(listen_addr))
+                            .map(|socket_addr| socket_addr.ip())
+                            .all(|listen_ip| listen_ip != observed_ip)
+                        {
+                            // TODO: Choose a random port from current local listen addresses
+                            if let Some(new_listen_addr) = self
+                                .listen_addrs
+                                .iter()
+                                .next()
+                                .and_then(|listen_addr| multiaddr_to_socketaddr(listen_addr))
+                                .map(|socket_addr| socket_addr.port())
+                                .map(|listen_port| {
+                                    addr.iter()
+                                        .map(|proto| match proto {
+                                            Protocol::Tcp(_) if info.session.ty.is_client() => {
+                                                // Replace only when it is an inbound connnection
+                                                Protocol::Tcp(listen_port)
+                                            }
+                                            value => value,
+                                        })
+                                        .collect()
+                                })
+                            {
+                                // TODO how can we trust this address?
+                                self.listen_addrs.push(new_listen_addr);
+                                if self
+                                    .addr_mgr
+                                    .add_observed_addr(&info.peer_id, addr.clone())
```

Comment:
In fact, the identity protocol only needs to report remote peer observed address by `add_observed_addr ` or events.

It's better to move all observe logic into the network crate, current implementation store listened addresses multiple time in different components which is unnecessary.

Reply:
We can't just store `listen_addrs` upper level. Because we should send updated `listen_addrs` to remote.

It's a chaos here, will be a big refactor later.

Reply:
The protocol can get updated `listen_addrs` from `addr_mgr` trait.

Reply:
I prefer refactor network first.